### ISSUE TITLE
Some queries slower after upgrade to 3.5.9

### ISSUE THREAD

**Issue by tomswinkels** [neo4j/neo4j#12286]
After upgrade to 3.5.9 some queries running slower.

In the release notes there are notes that there are something with indexes are fixed, but i think in some cases it is slower.

- Neo4j version: 3.5.9
- Operating system: Ubuntu 18.04
- API/Driver: Cypher

Query:

```
PROFILE
MATCH (startleaf:Hour{hash: '2019/08/19/07'}) , (endleaf:Hour{hash: '2019/09/03/07'}) , p = shortestPath((startleaf)-[:NEXT*0..]->(endleaf)) UNWIND nodes(p) AS leaf 
MATCH (leaf)<-[:SENDED]-(message:TS_P2000Message) 
WITH message AS message 
MATCH (message)-[:HAS_WORD]->(:TS_Word{name: 'amsterdam'}) 
WITH count(distinct message) AS results, collect(distinct message) AS messages 
WITH results AS results, messages AS messages 
UNWIND(messages) AS message 
WITH results, message AS message 
ORDER BY message.sended DESC 
SKIP 0 LIMIT 15 
RETURN results, message 
```

![plan-12](https://user-images.githubusercontent.com/9036151/64153338-f114cb00-ce2e-11e9-9bc8-5ebed9160a89.png)

**Comment by sherfert** [comment]
Maybe related: https://github.com/neo4j/neo4j/issues/12225

**Comment by tomswinkels** [comment]
@sherfert maybe.

First i was use 3.5.8 and the queries are very good. The message table has 6 mil records, but i has delete 3 mil records. After that i run the queries and they are very slow. But i has delete the half of the records.

After asking in the Neo4j Slack they say recreate all the indexes, after that, the queries are even slow. Now i have upgrade Neo4j to 3.5.9 and it is still the same problem but now a lot bigger, the query are so much slower.

It is weird that the queries are slower with the half of data.

**Comment by tomswinkels** [comment]
Today i have delete all the archived data but it still does not work....

The TS_P2000Message table had now a count of 975182.

**Comment by jexp** [comment]
can you add 

MATCH (message)-[:HAS_WORD]->(:TS_Word{name: 'amsterdam'}) 
USING INDEX ts:TS_Word(name)

even better


```
PROFILE
MATCH (word:TS_Word{name: 'amsterdam'}), (startleaf:Hour{hash: '2019/08/19/07'}) , (endleaf:Hour{hash: '2019/09/03/07'})
MATCH p = shortestPath((startleaf)-[:NEXT*0..]->(endleaf)) UNWIND nodes(p) AS leaf 
MATCH (leaf)<-[:SENDED]-(message:TS_P2000Message) 

WITH message 
WHERE exists( (message)-[:HAS_WORD]->(word)  )

WITH count(distinct message) AS results, collect(distinct message) AS messages 
WITH results AS results, messages AS messages 
UNWIND(messages) AS message 
WITH results, message AS message 
ORDER BY message.sended DESC 
SKIP 0 LIMIT 15 
RETURN results, message 

```

**Comment by tomswinkels** [comment]
@jexp thnx. That querie is faster.

But why my query is slower with deleting the half of my data? And after upgrade to the latest version of Neo4j again slower?

If i will match more words i need to do this?

```
PROFILE
MATCH (word1:TS_Word{name: 'someren'}), (word2:TS_Word{name: 'kerkstraat'}), (startleaf:Hour{hash: '2019/08/19/07'}) , (endleaf:Hour{hash: '2019/09/03/07'})
MATCH p = shortestPath((startleaf)-[:NEXT*0..]->(endleaf)) UNWIND nodes(p) AS leaf 
MATCH (leaf)<-[:SENDED]-(message:TS_P2000Message) 

WITH message 
WHERE 
exists( (message)-[:HAS_WORD]->(word1)  )
AND
exists( (message)-[:HAS_WORD]->(word2)  )
WITH count(distinct message) AS results, collect(distinct message) AS messages 
WITH results AS results, messages AS messages 
UNWIND(messages) AS message 
WITH results, message AS message 
ORDER BY message.sended DESC 
SKIP 0 LIMIT 15 
RETURN results, message
```

**Comment by jexp** [comment]
Could be a bug in index selectivity 

Can you share which labels went from which count to which?

**Comment by tomswinkels** [comment]
Did you mean the count of the nodes?

`MATCH (n:TS_P2000Message) RETURN count(n)`
978481
`MATCH (n:TS_Word) RETURN count(n)`
333834

**Comment by sherfert** [comment]
@tomswinkels If you could execute the following, in the state before and after deleting half the data, and share the output here that would help us investigate.

```
curl -H accept:application/json -H content-type:application/json -d '{"statements":[{"statement":"CALL db.stats.retrieve(\"GRAPH COUNTS\")"}]}' http://_________:7474/db/data/transaction/commit > graphCounts.json
```

This is all the statistics relevant to the planner.

**Comment by tomswinkels** [comment]
@sherfert here my CSV output:

[export-13.txt](https://github.com/neo4j/neo4j/files/3609090/export-13.txt)

**Comment by inmost-light** [comment]
@tomswinkels Could you try upgrading to a newer version? We think this issue might have been fixed in 3.5.12.

**Comment by tomswinkels** [comment]
@inmost-light I have the newest version of Neo4j 3.5 and i dont have anymore issues. But i have changed a lot, the whole database are now in the ram. I dont no if the bug is fixed in the version 3.5.12, but for now close this issue.
